In [1]:
import os
import pandas as pd
import copy
import numpy as np
from sklearn.linear_model import LinearRegression

# На основании данных рейтинга top500 спрогнозировать производительность компьютера №1 в 2025

In [2]:
file = pd.read_csv('top500/june 1993.csv', header=None)


In [3]:
file.head()

0          1              2       3        4                    5   \
0  NaN       Name  Manufacturer:  Cores:  Memory:           Processor:   
1  0.0  CM-5/1024            TMC   1,024     0 GB  SuperSPARC I  32MHz   
2  1.0   CM-5/544            TMC     544     0 GB  SuperSPARC I  32MHz   
3  2.0   CM-5/512            TMC     512     0 GB  SuperSPARC I  32MHz   
4  3.0   CM-5/512            TMC     512     0 GB  SuperSPARC I  32MHz   

                6                           7                         8   \
0    Interconnect:  Linpack Performance (Rmax)  Theoretical Peak (Rpeak)   
1  Hypercube, tree              0.0597 TFlop/s             0.131 TFlop/s   
2  Hypercube, tree              0.0304 TFlop/s           0.06963 TFlop/s   
3  Hypercube, tree              0.0304 TFlop/s           0.06554 TFlop/s   
4  Hypercube, tree              0.0304 TFlop/s           0.06554 TFlop/s   

       9               10                        11                 12  \
0    Nmax  HPCG [TFlop/s]  Power Measurement Level:  Operating System:   
1  52,224           False                     False              CMOST   
2  36,864           False                     False              CMOST   
3  36,864           False                     False              CMOST   
4  36,864           False                     False              CMOST   

          13             14     15      16      17  
0  Compiler:  Math Library:   MPI:  Power:   Nhalf  
1      False          False  False   False  24,064  
2      False          False  False   False  16,384  
3      False          False  False   False  16,384  
4      False          False  False   False  16,384

In [4]:
gen_table = pd.DataFrame()

In [5]:
for name in os.listdir('top500'):
    month = {'November': 11, 'December': 12, 'June': 6}
    tmp_file = pd.read_csv('top500/'+ name, header=None)
    tmp_file['month_in_progress'] =int((month[name.split(" ")[0]])-6 + 12*(int(name.split(" ")[1].split(".")[0]) - 1993))*30
    gen_table = gen_table.append(tmp_file.iloc[[1]], ignore_index=True)

In [6]:
gen_table.head()

0                      1        2      3     4                    5  \
0  0.0  Numerical Wind Tunnel  Fujitsu    140  0 GB      Fujitsu  105MHz   
1  0.0              CM-5/1024      TMC  1,024  0 GB  SuperSPARC I  32MHz   
2  0.0                XP/S140    Intel  3,680  0 GB   Intel 80860  50MHz   
3  0.0  Numerical Wind Tunnel  Fujitsu    140  0 GB      Fujitsu  105MHz   
4  0.0            SR2201/1024  Hitachi  1,024  0 GB      HARP-1E  150MHz   

                           6                7                8        9  \
0  Full distributed crossbar  0.19198 TFlop/s  0.23579 TFlop/s    False   
1            Hypercube, tree   0.0597 TFlop/s    0.131 TFlop/s   52,224   
2           2-D mesh (torus)   0.1434 TFlop/s    0.184 TFlop/s   55,700   
3  Full distributed crossbar  0.19198 TFlop/s  0.23579 TFlop/s    False   
4             Hyper crossbar   0.2324 TFlop/s   0.3072 TFlop/s  155,520   

      10     11         12     13     14     15     16      17  \
0  False  False      UXP/V  False  False  False  False   False   
1  False  False      CMOST  False  False  False  False  24,064   
2  False  False      OSF/1  False  False  False  False  20,500   
3  False  False      UXP/V  False  False  False  False   False   
4  False  False  HI-UX/MPP  False  False  False  False  34,560   

   month_in_progress  
0                900  
1                  0  
2                360  
3                720  
4               1080

In [7]:
gen_table = gen_table.iloc[:, [7, 18]]

In [8]:
gen_table.iloc[0, 0]

'0.19198 TFlop/s'

In [10]:
for i in range(57):
    a = copy.deepcopy(gen_table.iloc[i, 0].split(" ")[0])
    a = a.replace(",", "")
    a = float(a)

    gen_table.iloc[i, 0] = np.log(a*1000)

In [11]:
gen_table = gen_table.sort_values('month_in_progress')

In [13]:
predict = pd.read_csv('2025.csv', header=None)

In [14]:
predict[[0]]= 389*30

In [15]:
lr = LinearRegression()
lr.fit(np.array(gen_table['month_in_progress']).reshape(-1, 1), gen_table[7])

LinearRegression()

In [16]:
answer = lr.predict(predict[[0]])

In [18]:
np.exp(answer[0])

19827791151.224625

Прогнозируемая производительность топ-1 суперкомпьютера к 2025 году - 19.8 EFlop/s, хотя визуальный анализ графиков подсказывает, что к 2025 году производительность топ-1 суперкомпьютера будет лежать в границах 1.5-3 EFlop/s.

# Оцените какой процент всего производимого на Земле электричества потребляют все компьютеры из top500 

По данным British Petroleum в 2020 году было произведено 26823,248 ТВт*ч электроэнрегии

In [204]:
all_power = 26823.248

In [192]:
top500 = pd.read_csv('./Top500.csv', delimiter=';')

In [193]:
top500.columns.values

array(['Rank', 'Previous Rank', 'First Appearance', 'First Rank', 'Name',
       'Computer', 'Site', 'Manufacturer', 'Country', 'Year', 'Segment',
       'Total Cores', 'Accelerator/Co-Processor Cores', 'Rmax [TFlop/s]',
       'Rpeak [TFlop/s]', 'Nmax', 'Nhalf', 'HPCG [TFlop/s]', 'Power (kW)',
       'Power Source', 'Power Efficiency [GFlops/Watts]', 'Memory',
       'Architecture', 'Processor', 'Processor Technology',
       'Processor Speed (MHz)', 'Operating System', 'OS Family',
       'Accelerator/Co-Processor', 'Cores per Socket',
       'Processor Generation', 'System Model', 'System Family',
       'Interconnect Family', 'Interconnect', 'Continent', 'Site ID',
       'System ID'], dtype=object)

In [194]:
top500 = top500[['Power (kW)']]

In [195]:
top500

Power (kW)
0    29 899,23
1    10 096,00
2     7 438,28
3    15 371,00
4     2 528,00
..         ...
495        NaN
496     603,40
497        NaN
498        NaN
499        NaN

[500 rows x 1 columns]

In [196]:
top500.iloc[0, 0]

'29\xa0899,23'

In [197]:
top500['Power (kW)'] = top500['Power (kW)'].str.replace(',', '.').str.replace(u'\xa0', '').astype(float)

In [198]:
top500=top500['Power (kW)']

In [199]:
top500 = top500.interpolate(method ='linear', limit_direction ='backward')

In [200]:
top500

0      29899.230000
1      10096.000000
2       7438.280000
3      15371.000000
4       2528.000000
           ...     
495      643.022222
496      603.400000
497             NaN
498             NaN
499             NaN
Name: Power (kW), Length: 500, dtype: float64

In [201]:
top500_power = (top500.sum() * 8760) / 1e9
top500_power

7.5310469418

In [205]:
(top500_power / all_power) * 100

0.028076566051210502

Суперкомпьютеры из списка ТОП-500 за 6.2021 потребляют 0,028% всей произведенной электроэнергии в 2020 году